Module reload issue fix

keyonjie opened the pull request:
There is race between sof driver removing and hda codec jack-detecting
    work, which will lead to page fault issue as the latter work is
    accessing codec device which could be already removed.
    
    Here add cancellation of jack-detecting work before codecs are removed
    to avoid the race and fix the issue.
    
   This is aim to fix https://github.com/thesofproject/linux/issues/1067

ranj063 commented:
@keyonjie i guess the intention is correct here but I am struggling to understand why the SOF device has to explicitly do this. Shouldnt the codec device handle this when it is removed?Makes me wonder if something is amiss in the codec driver maybe?

plbossart commented:
> @keyonjie i guess the intention is correct here but I am struggling to understand why the SOF device has to explicitly do this. Shouldnt the codec device handle this when it is removed?Makes me wonder if something is amiss in the codec driver maybe?

I'm with Ranjani on this one.

keyonjie commented:
@ranj063 @plbossart Here it is actually the .remove() of our codec if you look into the details here:
```
hda_codec_driver_remove() {
	struct hda_codec *codec = dev_to_hda_codec(dev);

	if (codec->bus->core.ext_ops) {
		if (WARN_ON(!codec->bus->core.ext_ops->hdev_detach))
			return -EINVAL;
		return codec->bus->core.ext_ops->hdev_detach(&codec->core); //Keyon: So let's cancel the jack detection work for asoc codec driver here?
	}

	if (codec->patch_ops.free)
		codec->patch_ops.free(codec);
	snd_hda_codec_cleanup_for_unbind(codec); //Keyon: the jack detection work is cancelled here for legacy hda driver.
	module_put(dev->driver->owner);
	return 0;
}
```

xiulipan commented:
SOFCI TEST

plbossart commented:
Thanks @keyonjie for the pointer, it's hard to follow indirections that aren't part of the diffs.
I still don't understand what this code does in the first place. It's my understanding that we are using the 'patch' capabilities for HDaudio codecs, so we are obviously not releasing them. and then  I wonder if we also should call snd_hda_codec_cleanup_for_unbind(codec) for the ASoC case as well?

plbossart commented:
And I forgot to add that this jack detection work is not specific to ASoC so wondering why we need to do something here, we'd need to understand where the legacy driver does cancel the jack detection work.

ranj063 commented:
@keyonjie @plbossart I did some experiments today and the problem arises because of 		hda_codec_jack_check() call in hda_resume(). When the SOF device is removed during module unload, it is resumed from runtime suspend which causes the jackpoll work for all codecs to be scheduled. So it should be cancelled when the codec device is removed. 

@keyonjie I see that you've changed the PR to add that for the hda codec device. Do we need to do this for the hdmi codec device as well?

keyonjie commented:
Yes, as @ranj063 mentioned, we added scheduling the work in SOF driver hda_resume(), that's why we need this specific release here.

@ranj063 I first added it for all hda/hdmi codec in hda-codec.c, but when I talked this with @RanderWang, went into detail and found we only call snd_hda_codec_device_new()(where we initialize the delayed work codec->jackpoll_work) in hdac_hda.c(not in hdac_hdmi.c) only, then I switched to do this in hdac_hda.c only, I think it's safe enough at the moment, no need to do the same thing at the moment.

On the other hand, I remember @kv2019i had been investigating reusing legacy patch_hdmi.c driver in hdac_hdmi.c, once this transition done(snd_hda_codec_device_new() called there), we can add similar cancel_work in its .hdev_detach() ops.

What do you think? @plbossart @ranj063

keyonjie commented:
> Thanks @keyonjie for the pointer, it's hard to follow indirections that aren't part of the diffs.
> I still don't understand what this code does in the first place. It's my understanding that we are using the 'patch' capabilities for HDaudio codecs, so we are obviously not releasing them. and then I wonder if we also should call snd_hda_codec_cleanup_for_unbind(codec) for the ASoC case as well?

I thought of this, I think we should really revisit it carefully and split what are done in snd_hda_codec_cleanup_for_unbind() into groups of need-to-be-done and no-needed for asoc. 
I tried calling snd_hda_codec_cleanup_for_unbind() directly, but it looks too heavy for us, crashes happened with it.

keyonjie commented:
> And I forgot to add that this jack detection work is not specific to ASoC so wondering why we need to do something here, we'd need to understand where the legacy driver does cancel the jack detection work.

In snd_hda_codec_cleanup_for_unbind().

ranj063 commented:
@keyonjie the module unload/reload test for CML is horribly broken with all kinds of errors even if the jackpoll_work issue is fixed. we need to look into it

keyonjie commented:
> @keyonjie the module unload/reload test for CML is horribly broken with all kinds of errors even if the jackpoll_work issue is fixed. we need to look into it

What kinds of errors? Do you mean #1100 and #1101?

ranj063 commented:
> What kinds of errors? Do you mean #1100 and #1101?

@keyonjie no i didnt encounter these errors. But I suppose something was wrong with my branch. WIth #1026 and this PR things are looking stable with sof-dev. 

But in retrospect I think the logic to schedule the jackpoll_work for all codecs irrespective of whether there was a jack event or not was a bad idea I think. If we changed the logic in hda_codec_jack_check() to only schedule the jackpoll_work() if the event was from that codec, we'd never have this problem. What do you think?

keyonjie commented:
> But in retrospect I think the logic to schedule the jackpoll_work for all codecs irrespective of whether there was a jack event or not was a bad idea I think. If we changed the logic in hda_codec_jack_check() to only schedule the jackpoll_work() if the event was from that codec, we'd never have this problem. What do you think?

We already have check 'if (codec->jacktbl.used)', but here we are not scheduling work because of any event, instead we are doing that to detect any subsequent jack plugging/unplugging, in case the codec is supporting jack detection('if (codec->jacktbl.used)').

ranj063 commented:
> We already have check 'if (codec->jacktbl.used)', but here we are not scheduling work because of any event, instead we are doing that to detect any subsequent jack plugging/unplugging, in case the codec is supporting jack detection('if (codec->jacktbl.used)').

@keyonjie yes this is what I'm thinking might be flawed. Why should we do this? Rander's original PR only scheduled the jackdetection work if the status register bit for the codec was enabled.

keyonjie commented:
@ranj063 so are you meaning something like this:
```
	        status = azx_readw(chip, STATESTS);

		list_for_each_entry(codec, &bus->codec_list, list)
			if (status & (1 << codec->addr))
				queue_delayed_work(codec->bus->workq,
						   &codec->jackpoll_work, codec->jackpoll_interval);

```

Yes, I think this is better. But even with this, we still need to cancel the work at device removal.

ranj063 commented:
> Yes, I think this is better. But even with this, we still need to cancel the work at device removal.

@keyonjie but there wont be any work scheduled in the first place should be OK even if we dont but probably should to be safe. 
Besides one more thing with the current way things are is that the removal of sof_pci_dev takes a few seconds because of the work cancellation. have you noticed that too?

keyonjie commented:
> > Yes, I think this is better. But even with this, we still need to cancel the work at device removal.
> 
> @keyonjie but there wont be any work scheduled in the first place should be OK even if we dont but probably should to be safe.
> Besides one more thing with the current way things are is that the removal of sof_pci_dev takes a few seconds because of the work cancellation. have you noticed that too?

Yes, I observed that, it actually is the realtek codec runtime resume(runtime suspended->runtime resumed->driver removal) cost long time for initializing and regmap syncing before we can perform the real driver removal.

ranj063 commented:
@keyonjie I think we should make the change to schedule the jackpoll work only for the codec whose status bit is set. Also, the commit title and commit message needs updating for this PR as it is.

keyonjie commented:
> @keyonjie I think we should make the change to schedule the jackpoll work only for the codec whose status bit is set. 

Agree, but I am a little bit concern about the validation effort about jack detection feature itself, as some platforms might be already gone with the existed release.

> Also, the commit title and commit message needs updating for this PR as it is.

Do you mean about the PR subject/title or the commit message itself, what kind of update does it need?

ranj063 commented:
> Do you mean about the PR subject/title or the commit message itself, what kind of update does it need?

ASoC: SOF: Intel: hda: fix page fault issue via removing race

The commit title is incorrect as the changes are in ASoC: codecs: hda. 

The commit message itself needs to be updated too. The error is seen with the SOF driver but is not restricted to just SOF. In general for the ASoC case, the jackpoll work should be cancelled when the codec device is removed.

ranj063 commented:
> Agree, but I am a little bit concern about the validation effort about jack detection feature itself, as some platforms might be already gone with the existed release.
> 
> >
@keyonjie I think we need to make the change. This will affect the resume times and I'm sure at some point someone will point it out.

keyonjie commented:
> > Do you mean about the PR subject/title or the commit message itself, what kind of update does it need?
> 
> ASoC: SOF: Intel: hda: fix page fault issue via removing race
> 
> The commit title is incorrect as the changes are in ASoC: codecs: hda.
> 
> The commit message itself needs to be updated too. The error is seen with the SOF driver but is not restricted to just SOF. In general for the ASoC case, the jackpoll work should be cancelled when the codec device is removed.

Good catches, thanks, I wrote this with original patches version, which was add the cancel helpers in sof/intel/hda-codec.c, let me update them.

keyonjie commented:
> > Agree, but I am a little bit concern about the validation effort about jack detection feature itself, as some platforms might be already gone with the existed release.
> > >
> 
> @keyonjie I think we need to make the change. This will affect the resume times and I'm sure at some point someone will point it out.

OK, would you mind to change that in another PR/commit? That it is some change to Rander's original version in sof/intel/hda-codec.c, right?

keyonjie commented:
@ranj063 updated.

ranj063 commented:
@fredoh9 @xiulipan I do not see the ICL failure on the other PR's. Anything to be worried about before merging this PR?

wenqingfu commented:
> @fredoh9 @xiulipan I do not see the ICL failure on the other PR's. Anything to be worried about before merging this PR?

it looks like a CI test case stability issue.  You can go ahead to merge. We will look into it on Monday.
